### ISSUE TITLE
No connection pool... with multiple dbs and attributes 

### ISSUE THREAD

**Issue by jacobat** [rails/rails#36139]
### Steps to reproduce
Setup a new Rails 6.0.0rc1 app with a database config with two connections:

```yml
default: &default
  adapter: postgresql
  encoding: unicode
  pool: <%= ENV.fetch("RAILS_MAX_THREADS") { 5 } %>

development:
  primary:
    <<: *default
    database: multi_db_bug_development
  secondary:
    <<: *default
    database: multi_db_bug_development
```

Add a model that uses the connections with the new connection API. And also set an `attribute` on the model:

```ruby
class User < ApplicationRecord
  connects_to database: { foo: :primary, bar: :secondary }
  attribute :last_login_at, :datetime
end
```

Now open a Rails console and try to count the users:

```
irb(main):001:0> User.connected_to(role: :foo) do User.count end
Traceback (most recent call last):
        2: from (irb):1
        1: from (irb):1:in `block in irb_binding'
ActiveRecord::ConnectionNotEstablished (No connection pool with 'primary' found for the 'foo' role.)
irb(main):002:0> User.connected_to(role: :foo) do User.count end
   (1.4ms)  SELECT COUNT(*) FROM "users"
=> 0
```

### Expected behavior
The resulting value of `0` should returned on the first try.

### Actual behavior
First we get an exception. Following attempts works as expected.

### Interesting observations
Changing the roles from `foo` and `bar` to `writing` and `reading` makes the problem go away.

```
irb(main):001:0> User.connected_to(role: :writing) do User.count end
   (0.8ms)  SELECT COUNT(*) FROM "users"
=> 0
```

Removing the `attribute` statement from the model also makes the problem go away.

### Example app
A plain Rails 6.0.0rc1 app with the setup explained above is available at https://github.com/jacobat/rails-multi-db-bug

### System configuration
**Rails version**: 6.0.0rc1

**Ruby version**: 2.6.3

**Comment by y-yagi** [comment]
cc @eileencodes

**Comment by itsWill** [comment]
Hey @jacobat, I had a look at this and I think I have a somewhat 👋 hand-wavy 👋  idea of what is going on: 

1. Since we lazily define the attributes during the first call we have load the `@columns_hash` [here](https://github.com/rails/rails/blob/master/activerecord/lib/active_record/model_schema.rb#L487) 
2. Since we have an attribute we end up calling the attribute decorator to define it [here](https://github.com/rails/rails/blob/master/activerecord/lib/active_record/attribute_decorators.rb#L50)
3. This ends up calling the `load_schema` method on `attributes` [here](https://github.com/rails/rails/blob/master/activerecord/lib/active_record/attributes.rb#L246)
4. This in turn calls the `lookup` method, and this line is where the problem is: 

https://github.com/rails/rails/blob/ac1ba44f8d46ebbbe6889629d263e5690631f6a1/activerecord/lib/active_record/attributes.rb#L250

Since we don't pass in the `adapter_name` in the method, the lookup method triggers this call: 
https://github.com/rails/rails/blob/ac1ba44f8d46ebbbe6889629d263e5690631f6a1/activerecord/lib/active_record/type.rb#L52

Since we connect to the `ActiveRecord::Base.connection` the `connection_specification_name` defaults to `"primary"` but the `connection_specification_name` we're looking for is `"User"`.  This causes this [line]( https://github.com/rails/rails/blob/ac1ba44f8d46ebbbe6889629d263e5690631f6a1/activerecord/lib/active_record/connection_adapters/abstract/connection_pool.rb#L1018) to raise an error. 

 We can check that that is the case by doing something like `connection.send(:owner_to_pool).keys` before we hit the `load_schema` method for the first time to show that there is in fact no pool for `primary`. 

If we manually set the `@connection_specification_name` to `"User"` we'll see the expected behaviour. 

Now during the second call the `@columns_hash` is loaded and has values within it, that means that we don't have to hit the `lookup` code path, and this results in the second call succeeding since the `@connection_specification_name` stays as `"User"` through out. 

### What is the fix? 

We can avoid doing the `ActiveRecord::Base.connection.adapter_name` call by passing the adapter name of the current connection like so: 

``` patch
diff --git a/activerecord/lib/active_record/attributes.rb b/activerecord/lib/active_record/attributes.rb
index 7cf421c184..d6eca19c29 100644
--- a/activerecord/lib/active_record/attributes.rb
+++ b/activerecord/lib/active_record/attributes.rb
@@ -247,7 +247,7 @@ def load_schema! # :nodoc:
         super
         attributes_to_define_after_schema_loads.each do |name, (type, options)|
           if type.is_a?(Symbol)
-            type = ActiveRecord::Type.lookup(type, **options.except(:default))
+            type = ActiveRecord::Type.lookup(type, adapter: connection.adapter_name,  **options.except(:default))
           end
 
           define_attribute(name, type, **options.slice(:default))
```

The downside is I'm not sure the `connection` will always be defined, in which case we can guard with: 
``` ruby
connection.adapter_name if defined?(connection) 
``` 
I'm not sure of the case however where we'd have a `ActiveRecord::Type` without the `connection` defined, and making this change and running the `activerecord` tests they all pass happily. 

I'm happy to write up a PR for this with a test if the suggested approach is sound and you don't feel like doing it.

**Comment by jacobat** [comment]
Thanks for diving into this. I'm not qualified to speak on your discovery or suggested solution though so I'll leave that for others to do.

**Comment by eileencodes** [comment]
Looking at the bug report the problem here isn't the adapter name but the role names used.

Here you're using `foo` and `bar`

```
class User < ApplicationRecord
  connects_to database: { foo: :primary, bar: :secondary }
  attribute :last_login_at, :datetime
end
```

But the default in Rails is `writing` and `reading`. If you want to change them you need to change the default in your config/application.rb like this:

```
config.active_record.writing_role = :foo
config.active_record.reading_role = :bar
```

**Comment by eileencodes** [comment]
As a better explanation for why this is breaking:

Rails will always set up a `:writing` handler / role so if you want to use that existing connection you must either connect to the default (using `:writing` not `:foo`) or tell Rails you want to use a different default. Because Rails doesn't autoload in dev the foo role won't automatically be connected because we don't yet know about it. This is why it's important to tell Rails what the defaults are in your config/application.rb so that when Rails boots it connects to `:foo` instead of `:writing` and then everything will work correctly.

**Comment by jacobat** [comment]
Ah. I guess I was reading too much into this. So this feature is only for read/write splitting then? I was hoping to use it for basic sharding.

**Comment by eileencodes** [comment]
> Ah. I guess I was reading too much into this. So this feature is only for read/write splitting then? I was hoping to use it for basic sharding.

For now, yes. Rails connection handling was a mess and we had to do a lot of work just to get here. Hoping to get sharding in for 6.1 or as a bug fix in 6.0 if it doesn't require more massive refactoring.

**Comment by jacobat** [comment]
Alright. Thanks for the insight. I'll go ahead and close this as it isn't a bug then.
